File: app/src/main/java/com/ywwynm/everythingdone/model/ReminderHabitParams.java
package com.ywwynm.everythingdone.model;

import android.content.Context;

import com.ywwynm.everythingdone.App;
import com.ywwynm.everythingdone.utils.DateTimeUtil;

/**
 * Created by ywwynm on 2016/7/1.
 * params of reminder and habit
 */
public class ReminderHabitParams {

    private long   mReminderInMillis;
    private int[]  mReminderAfterTime;
    private int    mHabitType;
    private String mHabitDetail;

    public ReminderHabitParams() {
        mReminderInMillis  = -1;
        mReminderAfterTime = null;
        mHabitType         = -1;
        mHabitDetail       = null;
    }

    public ReminderHabitParams(ReminderHabitParams anotherParams) {
        mReminderInMillis = anotherParams.mReminderInMillis;
        if (anotherParams.mReminderAfterTime != null) {
            mReminderAfterTime = new int[2];
            System.arraycopy(anotherParams.mReminderAfterTime, 0,
                    mReminderAfterTime, 0, mReminderAfterTime.length);
        } else {
            mReminderAfterTime = null;
        }
        mHabitType   = anotherParams.mHabitType;
        mHabitDetail = anotherParams.mHabitDetail;
    }

    public long getReminderTime() {
        if (mReminderInMillis != -1) {
            return mReminderInMillis;
        }
        if (mReminderAfterTime != null) {
            return DateTimeUtil.getActualTimeAfterSomeTime(mReminderAfterTime);
        }
        return -1;
    }

    public void reset() {
        mReminderInMillis = -1;
        mReminderAfterTime = null;
        mHabitType = -1;
        mHabitDetail = null;
    }

    public String getDateTimeStr() {
        Context context = App.getApp();
        if (mReminderInMillis != -1) {
            return DateTimeUtil.getDateTimeStrAt(mReminderInMillis, context, false);
        } else if (mReminderAfterTime != null) {
            return DateTimeUtil.getDateTimeStrAfter(
                    mReminderAfterTime[0], mReminderAfterTime[1], context);
        } else {
            return DateTimeUtil.getDateTimeStrRec(context, mHabitType, mHabitDetail);
        }
    }

    public long getReminderInMillis() {
        return mReminderInMillis;
    }

    public void setReminderInMillis(long reminderInMillis) {
        this.mReminderInMillis = reminderInMillis;
    }

    public int[] getReminderAfterTime() {
        return mReminderAfterTime;
    }

    public void setReminderAfterTime(int[] reminderAfterTime) {
        this.mReminderAfterTime = reminderAfterTime;
    }

    public int getHabitType() {
        return mHabitType;
    }

    public void setHabitType(int habitType) {
        this.mHabitType = habitType;
    }

    public String getHabitDetail() {
        return mHabitDetail;
    }

    public void setHabitDetail(String habitDetail) {
        this.mHabitDetail = habitDetail;
    }


}

File: app/src/main/java/com/ywwynm/everythingdone/model/ReminderHabitParams.kt
package com.ywwynm.everythingdone.model

import com.ywwynm.everythingdone.App
import com.ywwynm.everythingdone.utils.DateTimeUtil

/**
 * Created by ywwynm on 2016/7/1.
 * Translated to Kotlin by ywwynm and Claude Opus 4.7 on 2026/5/20.
 * params of reminder and habit
 */
open class ReminderHabitParams() {

    var reminderInMillis: Long = -1
    var reminderAfterTime: IntArray? = null
    var habitType: Int = -1
    var habitDetail: String? = null

    constructor(anotherParams: ReminderHabitParams) : this() {
        reminderInMillis = anotherParams.reminderInMillis
        if (anotherParams.reminderAfterTime != null) {
            reminderAfterTime = IntArray(2)
            System.arraycopy(anotherParams.reminderAfterTime!!, 0,
                    reminderAfterTime!!, 0, reminderAfterTime!!.size)
        } else {
            reminderAfterTime = null
        }
        habitType = anotherParams.habitType
        habitDetail = anotherParams.habitDetail
    }

    open fun getReminderTime(): Long {
        if (reminderInMillis != -1L) {
            return reminderInMillis
        }
        if (reminderAfterTime != null) {
            return DateTimeUtil.getActualTimeAfterSomeTime(reminderAfterTime)
        }
        return -1
    }

    open fun reset() {
        reminderInMillis = -1
        reminderAfterTime = null
        habitType = -1
        habitDetail = null
    }

    open fun getDateTimeStr(): String? {
        val context = App.getApp()
        return if (reminderInMillis != -1L) {
            DateTimeUtil.getDateTimeStrAt(reminderInMillis, context, false)
        } else if (reminderAfterTime != null) {
            DateTimeUtil.getDateTimeStrAfter(
                reminderAfterTime!![0], reminderAfterTime!![1], context)
        } else {
            DateTimeUtil.getDateTimeStrRec(context, habitType, habitDetail)
        }
    }
}
